FWCore/Framework/src/Worker.h

```diff
@@ -441,7 +463,95 @@ namespace edm {
       typename T::Context const* m_context;
       ServiceToken m_serviceToken;
     };
-    
+
+    // AcquireTask is only used for the Event case, but we define
+    // it as a template so all cases will compile.
+    // DUMMY exists to work around the C++ Standard prohibition on
+    // fully specializing templates nested in other classes.
+    template <typename T, typename DUMMY = void>
+    class AcquireTask : public WaitingTask {
+    public:
+      AcquireTask(Worker* worker,
+                  typename T::MyPrincipal const& ep,
+                  EventSetup const& es,
+                  ParentContext const& parentContext,
+                  WaitingTaskWithArenaHolder holder) {}
+      tbb::task* execute() override { return nullptr; }
+    };
+
+    template <typename DUMMY>
+    class AcquireTask<OccurrenceTraits<EventPrincipal, BranchActionStreamBegin>, DUMMY> : public WaitingTask {
+    public:
+      AcquireTask(Worker* worker,
+                  EventPrincipal const& ep,
+                  EventSetup const& es,
+                  ParentContext const& parentContext,
+                  WaitingTaskWithArenaHolder holder):
+      m_worker(worker),
+      m_principal(ep),
+      m_es(es),
+      m_parentContext(parentContext),
+        m_holder(std::move(holder)),
+      m_serviceToken(ServiceRegistry::instance().presentToken()) {}
+
+      tbb::task* execute() override {
+        //Need to make the services available early so other services can see them
+        ServiceRegistry::Operate guard(m_serviceToken);
+
+        //incase the emit causes an exception, we need a memory location
+        // to hold the exception_ptr
+        std::exception_ptr temp_excptr;
+        auto excptr = exceptionPtr();
+        try {
+          //pre was called in prefetchAsync
+          m_worker->emitPostModuleEventPrefetchingSignal();
+        } catch(...) {
+          temp_excptr = std::current_exception();
+          if(not excptr) {
+            excptr = &temp_excptr;
+          }
+        }
+
+        if( not excptr) {
+          if(auto queue = m_worker->serializeRunModule()) {
+            Worker* worker = m_worker;
+            auto const & principal = m_principal;
+            auto& es = m_es;
+            auto parentContext = m_parentContext;
+            auto serviceToken = m_serviceToken;
+            queue.push( [worker, &principal, &es, parentContext, serviceToken, holder = m_holder]()
```

Comment:
If I may try to answer:

After execute is finished, the AcquireTask will be destroyed (as tbb::task is destroyed after execute has completed and there is no recycling).

So, you can not capture `this` directly as the local stack of the `execute` function will be gone but the module which is called is supposed to have access to some of these variables after execute is finished.

Vk

Reply:
Ah, I see, you cannot capture individual member variables, unless you assign them.
This would work:
```
[worker=m_worker, &principal=m_principal, &es=m_es, parentContext=m_parentContext, serviceToken=m_serviceToken, holder = m_holder](){...}
```
or even this if one wants to be confusing:
```
[m_worker=m_worker, &m_principal=m_principal, &m_es=m_es, m_parentContext=m_parentContext, m_serviceToken=m_serviceToken, m_holder = m_holder](){...}
```
but this does not:
```
[m_worker, &m_principal, &m_es, m_parentContext, m_serviceToken, m_holder](){...}

Thanks,
.A
```

Reply:
I agree with what was said above. This happened because I copy pasted this code from code written prior to C++14, so it originally had to use the C++11 syntax, then I added the last variable using the C++14 syntax. It would be better to be consistent and use the C++14 syntax throughout. I will change this in both places.